api/workers/events.go

```diff
@@ -0,0 +1,298 @@
+package workers
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/convox/rack/api/models"
+)
+
+var (
+	appLogStreams = map[string]appLogStream{}
+)
+
+type appLogStream struct {
+	LogGroup      string
+	LogStream     string
+	SequenceToken string
+}
+
+type queueHandler func(body string) error
+
+type event struct {
+	Account    string
+	DetailType string `json:"detail-type"`
+	Detail     interface{}
+	ID         string
+	Region     string
+	Resources  []string
+	Source     string
+	Time       time.Time
+	Version    string
+}
+
+type detailTaskStateChange struct {
+	ClusterArn string
+	Containers []struct {
+		ContainerArn string
+		LastStatus   string
+		Name         string
+		TaskArn      string
+	}
+	CreatedAt     time.Time
+	DesiredStatus string
+	Group         string
+	LastStatus    string
+	StartedAt     time.Time
+	StartedBy     string
+	StoppedReason string
+	TaskArn       string
+	UpdatedAt     time.Time
+}
+
+// StartEventQueue starts the event queue workers
+func StartEventQueue() {
+	go handleAccountEvents()
+	go handleCloudformationEvents()
+}
+
+func handleCloudformationEvents() {
+	err := processQueue("CloudformationEvents", func(body string) error {
+		var raw struct {
+			Message string
+			Subject string
+		}
+
+		if err := json.Unmarshal([]byte(body), &raw); err != nil {
+			return err
+		}
+
+		message := map[string]string{}
+
+		for _, line := range strings.Split(raw.Message, "\n") {
+			parts := strings.SplitN(line, "='", 2)
+
+			if len(parts) == 2 {
+				message[strings.TrimSpace(parts[0])] = strings.TrimSuffix(parts[1], "'")
+			}
+		}
+
+		app := strings.TrimPrefix(message["StackName"], fmt.Sprintf("%s-", os.Getenv("RACK")))
+
+		stream, err := getAppLogStream(app)
+		if err != nil {
+			return err
+		}
+
+		req := &cloudwatchlogs.PutLogEventsInput{
+			LogGroupName:  aws.String(stream.LogGroup),
+			LogStreamName: aws.String(stream.LogStream),
+		}
+
+		if stream.SequenceToken != "" {
+			req.SequenceToken = aws.String(stream.SequenceToken)
+		}
+
+		log := fmt.Sprintf("AWS/CloudFormation %s %s %s", message["ResourceStatus"], message["LogicalResourceId"], message["ResourceStatusReason"])
+
+		req.LogEvents = []*cloudwatchlogs.InputLogEvent{
+			&cloudwatchlogs.InputLogEvent{
+				Message:   aws.String(log),
+				Timestamp: aws.Int64(time.Now().UnixNano() / 1000000),
+			},
+		}
+
+		pres, err := models.CloudWatchLogs().PutLogEvents(req)
+		if err != nil {
```

Comment:
I believe this can error if you don't have the right sequence token. Generally you should handle that error specifically and grab a sequence token and retry